logservice.go

```diff
@@ -2,140 +2,177 @@ package main
 
 import (
 	"bufio"
-	"bytes"
 	"encoding/json"
 	"fmt"
+	"io"
+	"io/ioutil"
 	"log"
 	"os"
 	"path"
+	"sync"
 	"time"
 
-	"github.com/aws/aws-sdk-go/aws"
-	"github.com/aws/aws-sdk-go/aws/credentials"
-	"github.com/aws/aws-sdk-go/aws/session"
-	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/screwdriver-cd/log-service/s3fileuploader"
 )
 
 var emitterPath = "/var/run/sd/emitter"
 
-type logLine struct {
-	Time    int64  `json:"t"`
-	Message string `json:"m"`
-	Step    string `json:"s"`
-}
+const (
+	region         = "us-west-2"
+	bucket         = "logs.screwdriver.cd"
+	startupTimeout = time.Minute
+	uploadInterval = 2 * time.Second
+)
 
-type s3LogLine struct {
-	Time    int64  `json:"t"`
-	Message string `json:"m"`
-	Line    int    `json:"n"`
-}
+func main() {
+	if len(os.Args) != 2 {
+		log.Println("No buildID specified. Cannot log.")
+		os.Exit(0)
+	}
 
-func uploadLogToS3(content [][]byte, buildID, stepName string) {
-	keyid := os.Getenv("AWS_ACCESS_KEY_ID")
-	secretkey := os.Getenv("AWS_SECRET_ACCESS_KEY")
-	token := ""
-	creds := credentials.NewStaticCredentials(keyid, secretkey, token)
-	_, err := creds.Get()
+	buildID := os.Args[1]
+	log.Println("Processing logs for build", buildID)
+
+	// If we can't open the socket in the first 60s, the sender probably
+	// exited before transmitting any data. Since we are reading from
+	// a FIFO, we will block forever unless we bail.
+	t := time.AfterFunc(startupTimeout, func() {
+		log.Printf("No data in the first %s. Assuming catastophe.", startupTimeout)
+		os.Exit(0)
+	})
+	source, err := os.Open(emitterPath)
+	t.Stop()
 	if err != nil {
-		fmt.Printf("bad credentials: %s", err)
+		log.Printf("Failed opening %v: %v", emitterPath, err)
+		os.Exit(0)
+	}
+
+	uploader := s3fileuploader.NewS3FileUploader(region)
+	if err := archiveLogs(uploader, buildID, source); err != nil {
+		log.Printf("Error moving logs to S3: %v", err)
+		os.Exit(0)
 	}
+}
+
+// archiveLogs copies log lines from src into an S3 bucket.
+// Logs are copied to /buildID/stepName
+func archiveLogs(uploader s3fileuploader.S3FileUploader, buildID string, src io.Reader) error {
+	log.Println("Archiver started")
+
+	var wg sync.WaitGroup
+	steps := make(map[string]chan *logLine)
+	var lastStep string
 
-	cfg := aws.NewConfig().WithRegion("us-west-2").WithCredentials(creds)
-	svc := s3.New(session.New(), cfg)
+	scanner := bufio.NewScanner(src)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		newLog := &logLine{}
+		if err := json.Unmarshal([]byte(line), newLog); err != nil {
+			return fmt.Errorf("unmarshaling log line %s: %v", line, err)
+		}
 
-	combined := bytes.Join(content, []byte("\n"))
-	data := bytes.NewReader(combined)
-	size := int64(data.Len())
-	fmt.Println("Uploading this:", string(combined))
+		// Get a channel to send the log to
+		c, ok := steps[newLog.Step]
+		if !ok {
+			if steps[lastStep] != nil {
+				close(steps[lastStep])
+				delete(steps, lastStep)
```

Comment:
The working assumption is that steps will always be written sequentially. This will likely not always be the case if we decide to support backgrounding steps. We probably need to support events instead of just log lines at that point.